packages/ember-data/lib/serializers/rest-serializer.js

```diff
@@ -689,6 +689,49 @@ var RESTSerializer = JSONSerializer.extend({
     } else {
       json[key + "Type"] = Ember.String.camelize(belongsTo.modelName);
     }
+  },
+
+  /**
+    You can use this method to customize how a polymorphic relationship should
+    be extracted.
+
+    @method extractPolymorphicRelationship
+    @param {Object} relationshipModelName
+    @param {Object} relationshipHash
+    @param {Object} relationshipOptions
+    @return {Object}
+   */
+  extractPolymorphicRelationship: function(relationshipModelName, relationshipHash, relationshipOptions) {
+    var { relationshipKey, resourceHash, relationshipMeta } = relationshipOptions;
+
+    // A polymorphic belongsTo relationship can be present in the payload
+    // either in the form where the `id` and the `type` are given:
+    //
+    //   {
+    //     message: { id: 1, type: 'post' }
+    //   }
+    //
+    // or by the `id` and a `<relationship>Type` attribute:
+    //
+    //   {
+    //     message: 1,
+    //     messageType: 'post'
+    //   }
+    //
+    // The next code checks if the latter case is present and returns the
+    // corresponding JSON-API representation. The former case is handled within
+    // the super class (JSONSerializer).
+    var isPolymorphic = relationshipMeta.options.polymorphic;
+    var typeProperty = `${relationshipKey}Type`;
```

Comment:
`relationshipKey` is the result of `this.keyForRelationship(key, relationshipMeta.kind, 'deserialize');` from [this line](http://git.io/vCsx9).

In `serializePolymorphicType` above in line 686 `keyForAttribute` is used, so using `keyForAttribute` here would make it consistent I guess. But judging by the name, wouldn't be `keyForRelationship` the correct one, for both `serializePolymorphicType` and the new `serializePolymorphicType`?

It looks like AMS [uses `keyForRelationship](http://git.io/vCGTi) ...